src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/KeyValuePairModelBinder.cs

```diff
@@ -17,9 +17,13 @@ public sealed class KeyValuePairModelBinder<TKey, TValue> : IModelBinder
                 throw new ArgumentNullException(nameof(bindingContext));
             }
 
-            ModelBindingHelper.ValidateBindingContext(bindingContext,
-                                                      typeof(KeyValuePair<TKey, TValue>),
-                                                      allowNullModel: true);
+            if (bindingContext.ModelType != typeof(KeyValuePair<TKey, TValue>))
+            {
+                var message = Resources.FormatModelBinderUtil_ModelTypeIsWrong(
+                    bindingContext.ModelType,
+                    typeof(KeyValuePair<TKey, TValue>));
+                throw new ArgumentException(message, nameof(bindingContext));
+            }
```

Comment:
Hmm, depends on whether we think anything but `GenericModelBinder` will ever interact with this class. If this is only a `GenericModelBinder` toy, `Debug.Assert()` would be appropriate. Otherwise returning might make sense.

But I do agree `throw`ing seems un-model binder-ish.